### PR TITLE
refactor(ListItem): createListItemPartitions

### PR DESCRIPTION
**Pull request by ElliotTheHuman**
`createListItemPartitions` accepts all `ListItemProps` and outputs the core pieces that make up `ListItem` (i.e. the item, the label, the detail). It outputs these pieces in a 2-tuple, the first element being a fragment that contains all elements to go "inside" the main container, and a second element being all elements that go "outside" the main container (which is just the detail when accessory mode is enabled).

The PR does the following:
- Adds `createListItemPartitions`
- Refactors `ListItem` pieces to utilize `createListItemPartitions`

## Developer Checklist [ℹ️](https://github.com/looker-open-source/components/blob/main//CONTRIBUTING.md#developer-checklist)

- [ ] ♿️ Accessibility addressed
- [ ] 🌏 Localization & Internationalization addressed
- [ ] 🖼 Image Snapshot coverage
- [ ] 📚 Documentation updated
- [ ] 🧳 Bundle size impact addressed
- [ ] 🏁 Performance impacts addressed
- [ ] 👾 Browsers tested
  - [ ] Chrome
  - [ ] Firefox
  - [ ] Safari
  - [ ] IE11

### CODE REVIEW

**Comment by ElliotTheHuman** [comment]
Hey @dbchristopher! Tagging you in this PR as well to try and get a reviewer that was further away from these changes. Cheers.

File: packages/components/src/List/ListItem.tsx [hunk 1]
```diff
@@ -185,17 +134,13 @@ const ListItemInternal = forwardRef(
...
-          {content}
-        </ListItemDetail>
-      </HoverDisclosure>
-    )
+    const [ariaProps, wrapperProps] = partitionAriaProps(restProps)
+    const [insideElements, outsideElements] = useListItemPartitions({
+      ...props,
+      color,
+      density,
+      icon: icon || (iconGutter ? <IconPlaceholder /> : undefined),
```
Comment (lukelooker) [inline packages/components/src/List/ListItem.tsx:142]:
```suggestion
      icon: icon || (iconGutter && <IconPlaceholder />),
```

File: packages/components/src/List/ListItemLabel.tsx [hunk 2]
```diff
@@ -87,6 +93,7 @@ export const ListItemLabel = styled(ListItemLabelLayout).withConfig({
 })`
   ${({ height, itemRole }) => itemRole === 'none' && `min-height: ${height}px;`}
   ${listItemBackgroundColor}
+  ${(props) => listItemPadding({ ...props })}
```
Comment (lukelooker) [inline packages/components/src/List/ListItemLabel.tsx:96]: I have questions... I think.

File: packages/components/src/List/TruncateWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+export const TruncateWrapper: FC<{
+  color?: string
+  fontSize?: FontSizes
+  lineHeight?: FontSizes
+  truncateDescription?: string
+}> = ({ children, color, truncateDescription, fontSize, lineHeight }) => (
+  <Text color={color} fontSize={fontSize} lineHeight={lineHeight}>
+    <Truncate description={truncateDescription}>{children}</Truncate>
+  </Text>
```
Comment (lukelooker) [inline packages/components/src/List/TruncateWrapper.tsx:40]: I wonder if we should just admit defeat and support those props on `Truncate`?

File: packages/components/src/List/utils/useListItemPartitions.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import { Flex } from '../../Layout'
+import { listItemDimensions } from './listItemDimensions'
+import { listItemIconColor, listItemLabelColor } from './listItemColor'
+import { getDetailOptions } from '.'
+
+export type UseListItemPartitions = {
+  /**
+   * Should be either compatible with IconType or an <IconPlaceholder /> component
+   */
+  icon?: ReactNode
```
Comment (lukelooker) [inline packages/components/src/List/utils/useListItemPartitions.tsx:43]:
```suggestion
  icon?: IconType | ReactElement<IconPlaceholderProps>
```

File: packages/components/src/List/utils/useListItemPartitions.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  const renderedIcon = icon && (
+    <ListItemIcon
+      color={listItemIconColor(color, disabled)}
+      disabled={disabled}
+      iconGap={itemDimensions.iconGap}
+      iconSize={itemDimensions.iconSize}
+    >
+      {icon}
+    </ListItemIcon>
+  )
```
Comment (lukelooker) [inline packages/components/src/List/utils/useListItemPartitions.tsx:68]: If the hook is too long to read easily why not extract these as functions unto themselves:

```suggestion
  const renderedIcon = generateIcon(...)
```

```
const generateIcon = (icon, color, disabled, dimensions) => icon ? (
    <ListItemIcon
      color={listItemIconColor(color, disabled)}
      disabled={disabled}
      iconGap={dimensions.iconGap}
      iconSize={dimensions.iconSize}
    >
      {icon}
    </ListItemIcon>
  ) : null

File: packages/components/src/List/utils/useListItemPartitions.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      {description}
+    </Wrapper>
+  )
+
+  const renderedContent = (
+    <Flex flexDirection="column" minWidth={0} flexGrow={1}>
+      {labelElement}
+      {descriptionElement}
+    </Flex>
+  )
```
Comment (lukelooker) [inline packages/components/src/List/utils/useListItemPartitions.tsx:103]: Extract into function :)

File: packages/components/src/List/utils/useListItemPartitions.tsx [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        pr={accessory && padding ? itemDimensions.px : '0'}
+      >
+        {content}
+      </ListItemDetail>
+    </HoverDisclosure>
+  )
+
+  const inside = (
+    <>
+      {renderedIcon}
```
Comment (lukelooker) [inline packages/components/src/List/utils/useListItemPartitions.tsx:124]: Use extracted functions directly here:

```suggestion
      {generateIcon(icon, color, disabled, dimensions)}
```

**Review by lukelooker — CHANGES_REQUESTED** [review]
Ok - overall this looks great!

Suggestions on how to clean-up the readability of the hook and one open question that I'd like to do via 1:1

File: packages/components/src/List/TruncateWrapper.tsx (hunk 3 above)
Comment (lukelooker) [reply packages/components/src/List/TruncateWrapper.tsx:40]: Working on this now.

File: packages/components/src/List/ListItem.tsx (hunk 1 above)
Comment (lukelooker) [reply packages/components/src/List/ListItem.tsx:142]: Nevermind. I'm wrong.

File: packages/components/src/List/TruncateWrapper.tsx (hunk 3 above)
Comment (ElliotTheHuman) [reply packages/components/src/List/TruncateWrapper.tsx]: 🎉 🎉 🎉

File: packages/components/src/List/utils/useListItemPartitions.tsx (hunk 4 above)
Comment (ElliotTheHuman) [reply packages/components/src/List/utils/useListItemPartitions.tsx:42]: That's what I was looking for! Couldn't figure out the best way to do "React Component with X specific shape".

**Pull request closed without merging**